Валидация в формах отправки
---
---
Когда нам требуется специальная валидация, когда значение одного поля
должно проверяться в зависимости от другого поля, то нам будет недостаточно
валидации на уровне полей модели.

---

И так, представим ситуацию когда у нас есть модель которую мы подключаем в
административную панель, по дефолту на основании полей модели будут 
сгенерированы поля для формы.

```python
from django.db import models

class MyModel(models.Model):
    """Модель"""
    name = models.CharField('название')
    type = models.CharField('тип')

    def __str__(self):
        return str(self.pk)

    class Meta:
        verbose_name = 'Моя модель'
        verbose_name_plural = 'Мои модели'
```

Если нам требуется управлять формой, то в спец атрибуте `form` можно 
указать свою форму которая и будет использована. 

```python
from django.contrib import admin
from forms import MyModel

class DocBasketAdmin(admin.ModelAdmin):
    """Админ панель для сгенерированных комплектов документов"""
    list_display = (....)
    form = MyModel
```

Создаем свою собственную форму отправки `MyForm`, по дефолту если ничего не 
менять то все поля для формы так же будут сгенерированы на основе полей модели,
мы можем переопределить эти поля при помощи специальных виджетов.

```python
from django import forms

class MyForm(forms.ModelForm):

    class Meta:
        """Описание мета информации о классе"""
        model = MyModel
        fields = '__all__'
```

В процессе своего жизненного цикла форма проходит множество этапов, на этапе
когда поля формы проверены, но сам созданный объект модели еще не был сохранен,
объект формы проходит через вызов метода `clean`

```python
class BaseModelForm(BaseForm):
    ...
    def clean(self):
        self._validate_unique = True
        return self.cleaned_data
    ...
```

Официальная документация рекомендует именно на этапе валидации этого метода,
делать кастомную проверку, так что можно переопределить этот метод и добавить 
свою валидацию формы.

Для вызова исключений на конкретное поле формы, есть специальный метод самой 
формы, метод `add_error` принимает название поля и ошибку валидации, ошибка
должна быть класса `ValidationError`, подобная ошибка будет привязана именно
к тому полю у которого она вызвана.

```python
def clean(self):
    ...
    Валидация
    ...
    self.add_error(
        field='field_name',
        error=ValidationError(
            message='Не верные номера : %s',
            params='; '.join(_numbers)
            )
    )
    ...
```




